sys/net/gnrc/sock/gnrc_sock.c

```diff
@@ -162,6 +162,19 @@ ssize_t gnrc_sock_recv(gnrc_sock_reg_t *reg, gnrc_pktsnip_t **pkt_out,
 #elif IS_USED(MODULE_XTIMER)
     xtimer_remove(&timeout_timer);
 #endif
+
+#if IS_USED(MODULE_XTIMER) || IS_USED(MODULE_ZTIMER_USEC)
+    /* The timeout may have fired after the message was received but
+     * before the timer was removed. In this case, the mbox contains the
+     * expected message followed by the unexpected timeout message. We need
+     * to drain the unexpected timeout message to not have a bogus timeout
+     * error on the next call or confusing other places in the code */
+    {
+        msg_t garbage_bin;
+        mbox_try_get_with_type(&reg->mbox, &garbage_bin, _TIMEOUT_MSG_TYPE);
```

Comment:
I think this still has potential for racey behavior. What if two threads are waiting with timeout? Then this clears out the timeout of the other thread. IMHO, the most ideal way to fix this to use the PID of the waiter as a value for the timeout message instead of some magic integer and check here, if the removed timeout message is actually the one for this thread. Since this requires to search for both type and value, I am not sure, if having this dedicated function makes that much sense.

Reply:
Is it legal to have two threads waiting on the same socket? If so, whats the use case?

On mutli-core systems one could use that to distribute load on multiple cores. I don't see this happen on RIOT.